File: [0322][Medium][Coin_Change]/Coin_Change_2.py
# leetcode time     cost : 172 ms
# leetcode memory   cost : 13.7 MB 
# Time  Complexity: O(n)
# Space Complexity: O(n)
# slolution 2, Dynamic programming - Bottom up,loop money amount with recursion
class Solution:
  def coinChange(self, coins: List[int], amount: int) -> int:
    coins.sort(reverse = True)
    min_coins = float('inf')
    
    def count_coins(start_coin, coin_count, remaining_amount):
      nonlocal min_coins
      
      if remaining_amount == 0:
        min_coins = min(min_coins, coin_count)
        return
      
      # Iterate from largest coins to smallest coins
      for i in range(start_coin, len(coins)):
        remaining_coin_allowance = min_coins - coin_count
        max_amount_possible = coins[i] * remaining_coin_allowance
        
        if coins[i] <= remaining_amount and remaining_amount < max_amount_possible:
          count_coins(i, coin_count + 1, remaining_amount - coins[i])
      
    count_coins(0, 0, amount)
    return min_coins if min_coins < float('inf') else -1

File: [0322][Medium][Coin_Change]/Coin_Change_2.java
// leetcode time     cost : 43 ms
// leetcode memory   cost : 41 MB 
// Time  Complexity: O(S*n), S is the needed money amount
// Space Complexity: O(S)
// https://leetcode-cn.com/problems/coin-change/solution/ling-qian-dui-huan-by-leetcode/
// slolution 2, Dynamic programming - Top down,loop coin with recursion
public class Solution {

    public int coinChange(int[] coins, int amount) {
      if (amount < 1) return 0;
      return coinChange(coins, amount, new int[amount]);
    }
  
    private int coinChange(int[] coins, int rem, int[] count) {
      if (rem < 0) return -1;
      if (rem == 0) return 0;
      if (count[rem - 1] != 0) return count[rem - 1];
      int min = Integer.MAX_VALUE;
      // loop the all possible coin in the outer iteration,then do the deeper recursion.
      for (int coin : coins) {
        int res = coinChange(coins, rem - coin, count);
        if (res >= 0 && res < min)
          min = 1 + res;
      }
      count[rem - 1] = (min == Integer.MAX_VALUE) ? -1 : min;
      return count[rem - 1];
    }
}